src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+        /* unset lock */
+        flock(lockfd, LOCK_UN);
+        return PMIX_ERROR;
+    }
+
+    /* get ns_track_elem_t object for the target namespace from the local track list. */
+    elem = _get_track_elem_for_namespace(nspace);
+    if (NULL == elem) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        /* unset lock */
+        flock(lockfd, LOCK_UN);
+        return PMIX_ERROR;
+    }
+    /* need to update tracker:
+     * attach to shared memory regions for this namespace and store its info locally
+     * to operate with address and detach/unlink afterwards. */
+    rc = _update_ns_elem(elem, ns_info);
+    if (PMIX_SUCCESS != rc) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        /* unset lock */
+        flock(lockfd, LOCK_UN);
+        return PMIX_ERROR;
+    }
+
+    /* now we know info about meta segment for this namespace. */
+    meta_seg = elem->meta_seg;
+    data_seg = elem->data_seg;
+
+    while (nprocs--) {
+        if (PMIX_RANK_UNDEF == rank) {
+            cur_rank++;
+        }
+        /* Then we look for the rank meta info in the shared meta segment. */
+        rinfo = _get_rank_meta_info(cur_rank, meta_seg);
+        if (NULL == rinfo) {
+            PMIX_OUTPUT_VERBOSE((7, pmix_globals.debug_output,
+                        "%s:%d:%s:  no data for this rank is found in the shared memory. rank %d", __FILE__, __LINE__, __func__, cur_rank));
+            rc = PMIX_ERR_PROC_ENTRY_NOT_FOUND;
+            continue;
+        }
+        addr = _get_data_region_by_offset(data_seg, rinfo->offset);
+        if (NULL == addr) {
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            rc = PMIX_ERR_PROC_ENTRY_NOT_FOUND;
+            continue;
+        }
+        kval_cnt = rinfo->count;
+        /* TODO: probably PMIX_ERR_NOT_FOUND is a better way but
+         * setting to one initiates wrong next logic for unknown reason */
+        rc = PMIX_ERROR;
+
+        while (0 < kval_cnt) {
+            /* data is stored in the following format:
+             * key[PMIX_MAX_KEYLEN+1]
+             * size_t size
+             * byte buffer containing pmix_value, should be loaded to pmix_buffer_t and unpacked.
+             * next kval pair
+             * .....
+             * EXTENSION slot which has key = EXTENSION_SLOT and a size_t value for offset to next data address for this process.
+             */
+            if (0 == strncmp((const char *)addr, "INVALIDATED", PMIX_MAX_KEYLEN+1)) {
+                PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                            "%s:%d:%s: for rank %s:%d, skip INVALIDATED region", __FILE__, __LINE__, __func__, nspace, cur_rank));
+                /*skip it */
+                size_t size = *(size_t *)(addr + PMIX_MAX_KEYLEN + 1);
+                /* go to next item, updating address */
+                addr += KVAL_SIZE(size);
+            } else if (0 == strncmp((const char *)addr, "EXTENSION_SLOT", PMIX_MAX_KEYLEN+1)) {
```

Comment:
@igor-ivanov 
I suggest to define a constant instead of hardcoding "EXTENSION_SLOT"
Also something like "PMIX_INT_EXT_SLOT" would be safer in terms of the key name since we are using it in the user's key space and someone can easily use "EXTENSION_SLOT" as the name of the key.